### PR TITLE
Install script for Spin

### PR DESCRIPTION
**Pull request by mikkelhegn**
Signed-off-by: Mikkel Mørk Hegnhøj <mikkel@fermyon.com>

This is the first commit for an installer script for Spin, to make the acquisition story easier.

Current progress on testing:

- [x] Tested on MacBook with Apple Silicon
- [ ] Tested on MacBook with Intel
- [x] Tested on Linux with Intel
- [x] Tested on Linus with ARM processor (should fail)
- [x] Tested on WSL
- [x] Get the latest version from GitHub - it's currently hard-coded in the script
- [x] Support installing a specific version (including canary), through an argument (e.g., install.sh -v canary)

A subsequent PR could have the equivalent for PowerShell.

Relates to #641

### CODE REVIEW

**Comment by bacongobbler** [comment]
Works on WSL.

> * Get the latest version from GitHub - it's currently hard-coded in the script

This might be helpful: https://github.com/helm/helm/blob/8199db309ae73d745b983fb7ea2bfdd944a834ae/scripts/get-helm-3#L102-L115

File: install.sh [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    VERSION=$(wget -qO- https://github.com/fermyon/spin/releases | grep 'href="/fermyon/spin/releases/tag/v[0-9]*.[0-9]*.[0-9]*\"' | sed -E 's/.*\/fermyon\/spin\/releases\/tag\/(v[0-9\.]+)".*/\1/g' | head -1)
+fi
+
+# Constructing download FILE and URL
+FILE="spin-${VERSION}-${OS}-${ARC}.tar.gz"
+URL="https://github.com/fermyon/spin/releases/download/${VERSION}/${FILE}"
+
+# Download file, exit if not found - e.g. version does not exist
+fancy_print 0 "Step 1: Downloading: ${URL}"
+wget -q $URL || (fancy_print 1 "The request file does not exist: ${FILE}"; exit 1)
```
Comment (vdice) [inline install.sh:116]:
```suggestion
wget -q $URL || (fancy_print 1 "The requested file does not exist: ${FILE}"; exit 1)
```

File: install.sh [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+./spin --version
+fancy_print 0 "Done...\n"
+
+# Remove the compressed file
+fancy_print 0 "Step 3: Removing the downloaded tarball"
+rm $FILE
+fancy_print 0 "Done...\n"
+
+# Direct to quicks-start doc
+fancy_print 0 "You're good to go. Check here for for the next steps: https://spin.fermyon.dev/quickstart/"
```
Comment (vdice) [inline install.sh:131]:
```suggestion
fancy_print 0 "You're good to go. Check here for the next steps: https://spin.fermyon.dev/quickstart/"
```

File: install.sh [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    ;;
+*)
+    fancy_print 1 "The Processor: ${UNAME_ARC} is not supported by Spin."
+    exit 1
+    ;;
+esac
+
+case $OSTYPE in
+"linux-gnu"*)
+    OS="linux"
```
Comment (vdice) [inline install.sh:91]: Do we want to handle the specific case when OSTYPE is linux-gnu and ARC is aarch64 here?  eg print a line stating this combination isn't yet supported and exit 1?

**Review by vdice — COMMENTED** [review]
Looking good so far and works for me on my setup (Mac w/ Apple Silicon).  A few notes/Qs.

Is this PR still `WIP`?

Lastly, as an optional follow-up, what do we think about hosting this installation script on the Spin docs website?  Then users could (if comfortable) run eg `curl https://spin.fermyon.dev/install.sh | bash`, optionally with a specific `VERSION`.

File: install.sh [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+fancy_print 0 "Done...\n"
+
+# Remove the compressed file
+fancy_print 0 "Step 3: Removing the downloaded tarball"
+rm $FILE
+fancy_print 0 "Done...\n"
+
+# Direct to quicks-start doc
+fancy_print 0 "You're good to go. Check here for for the next steps: https://spin.fermyon.dev/quickstart/"
+fancy_print 0 "A good place to start is to install templates once you have all requirement."
```
[the remark this answers is not in the document]
Comment (mikkelhegn) [reply install.sh:132]: I've seen more people trying our Rust and then nt having the wasm32-wasi target installed. I'll try and rephrase to go and look for the language specific requirements, that really what I meant :-) Reverting on this one. The quickstart article should tell you the next steps, let's not make those redundant in this script.

File: install.sh (hunk 3 above)
Comment (mikkelhegn) [reply install.sh:91]: Oh - I see the combination of Linux and aarch64 is not caught as not-supported.

Comment (mikkelhegn) [reply install.sh:91]: uname -m on Mac return "arm64" on Linux it returns "aarch64", so this was originally caught when evaluating the processor type, but now I've made that case specific.

**Comment by mikkelhegn** [comment]
@vdice - the latest update has moved the script to /downloads on the docs site. The quickstart doc now describes how to use the script to get Spin.

Let me know what you think.

**Review by vdice — APPROVED** [review]
Looks and works great!  Noted a few last spots to update but otherwise LGTM.

**Comment by mikkelhegn** [comment]
Had to rebase and merge, hence just one commit now. @vdice let me know if everything still looks good, and feel free to merge.

**Review by vdice — COMMENTED** [review]
@mikkelhegn one last item:  It looks like the base README.md snuck in on a recent commit, duplicated in the downloads folder.  Let's remove this and then we're ready to go!

File: docs/downloads/install.sh [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    fancy_print 1 "The OSTYPE: ${OSTYPE} is not supported by this script."
+    fancy_print 2 "Please refer to this article to install Spin: https://spin.fermyon.dev/quickstart/"
+    exit 1
+    ;;
+esac
+
+# Check desired version. Default to latest if no desired version was requested
+if [[ $VERSION != "" ]]; then
+    :
+else
```
Comment (lann) [inline docs/downloads/install.sh:110]:
```suggestion
if [[ $VERSION = "" ]]; then
```

**Comment by mikkelhegn** [comment]
@vdice - let's see if that's a wrap 😂

File: docs/content/quickstart.md [hunk 6]
```diff
@@ -12,36 +12,49 @@ url = "https://github.com/fermyon/spin/blob/main/docs/content/quickstart.md"
...
 ```
 
-At this point, move the `spin` binary somewhere in your path, so it can be
-accessed from any directory.
+### Alternative 2: Using Cargo to install Spin
+
+If you have [`cargo`](https://doc.rust-lang.org/cargo/getting-started/installation.html), you can clone the repo and install it to your path:
+
+```console
+$ git clone https://github.com/fermyon/spin -b v0.4.0
```
Comment (vdice) [inline docs/content/quickstart.md:52]:
```suggestion
$ git clone https://github.com/fermyon/spin -b v0.5.0
```

**Review by vdice — COMMENTED** [review]
@mikkelhegn spotted one last little change  -- then we're good!

**Review by vdice — APPROVED** [review]

**Comment by gdamjan** [comment]
btw,
on github releases, the `/latest` url redirects to the latest release (and latest is the status of the release, not the newest).

this is the url for spin:
https://github.com/fermyon/spin/releases/latest

**Pull request merged**